### PR TITLE
MRG: allow users to annotate epochs with colors

### LINKED ISSUE #6737: visualization of autoreject results in MNE
Currently, the `autoreject` code hacks into MNE by copying the epochs plotting code and adding [two parameters](https://github.com/autoreject/autoreject/blob/master/autoreject/viz.py#L123). Unfortunately this is a [bit unwieldy to maintain](https://github.com/autoreject/autoreject/issues/153) on the `autoreject` side as we're using a ton of private functions in MNE and duplicating almost 1000 lines of code ... 

are people open to adding these parameters to `epochs.plot` in MNE, or alternatively some code to interpret the `drop_log` in a special way if it starts with the keyword `AUTOREJECT`. I can make a PR with the existing diff and iterate the API from there if people are okay with this.

### PR DESCRIPTION
**Pull request by jasmainak**
closes #6737 

just started by copying over changes from autoreject. Will iterate tomorrow

### CODE REVIEW

File: mne/viz/epochs.py [hunk 1]
```diff
@@ -812,6 +813,11 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
 
         .. versionadded:: 0.18.0
     %(show_scrollbars)s
+    bad_epochs_idx : array-like | None
+        Indices of bad epochs to show. No bad epochs to visualize if None.
+    log_labels : dataframe, shape (n_channels, n_epochs) | None
+        The bad segments to show in red and the interpolated segments
+        to show in green.
```
Comment (larsoner) [inline mne/viz/epochs.py]: This does not look like the API we discussed previously, it should just involve `epoch_colors` IIRC

https://github.com/mne-tools/mne-python/issues/6737#issuecomment-528440697

Comment (jasmainak) [reply mne/viz/epochs.py]: yes, I am just trying to get it to work first :)

**Comment by jasmainak** [comment]
okay, this now works in conjunction with this PR on autoreject: https://github.com/autoreject/autoreject/pull/155/files

next step is tests

File: mne/viz/epochs.py [hunk 2]
```diff
@@ -812,6 +812,8 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
 
         .. versionadded:: 0.18.0
     %(show_scrollbars)s
+    color : list of (n_epochs) list (of n_channels) | None
```
Comment (larsoner) [inline mne/viz/epochs.py]: list of list, shape (n_epochs, n_channels)

File: mne/viz/tests/test_epochs.py [hunk 3]
```diff
@@ -83,6 +83,17 @@ def test_plot_epochs_basic(capsys):
     with pytest.warns(RuntimeWarning, match='projection'):
         epochs.plot(noise_cov=cov)
     plt.close('all')
+    # check the epochs color plotting
+    color = [['r'] * len(epochs.ch_names) for _ in range(len(epochs.events))]
+    epochs.plot(color=color)
+    pytest.raises(TypeError, epochs.plot, color='r')
```
Comment (larsoner) [inline mne/viz/tests/test_epochs.py]: This is the old way, use the new one:
```
with pytest.raises(TypeError, match='color must be a list'):
    epochs.plot(color='r')
```

**Comment by jasmainak** [comment]
comments addressed

File: mne/viz/epochs.py [hunk 4]
```diff
@@ -808,6 +809,8 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
 
         .. versionadded:: 0.18.0
     %(show_scrollbars)s
+    color : list of list, shape (n_epochs, n_channels) | None
+        Colors to use for individual epochs. If None, use default colors
```
Comment (agramfort) [inline mne/viz/epochs.py]:
```suggestion
        Colors to use for individual epochs. If None, use default colors.
```

File: mne/viz/tests/test_epochs.py [hunk 5]
```diff
@@ -83,6 +83,19 @@ def test_plot_epochs_basic(capsys):
     with pytest.warns(RuntimeWarning, match='projection'):
         epochs.plot(noise_cov=cov)
     plt.close('all')
+    # check the epochs color plotting
+    color = [['r'] * len(epochs.ch_names) for _ in range(len(epochs.events))]
+    epochs.plot(color=color)
+    with pytest.raises(TypeError, match='must be an instance of'):
+        epochs.plot(color='r')
+    with pytest.raises(ValueError, match='color must be list'):
```
Comment (agramfort) [inline mne/viz/tests/test_epochs.py]:
```suggestion
    with pytest.raises(ValueError, match='must be list of len'):
```

**Comment by drammock** [comment]
@jasmainak would be great to have a demo of this functionality in one of the examples / tutorials.

File: mne/viz/epochs.py [hunk 6]
```diff
@@ -978,6 +982,22 @@ def _prepare_mne_browse_epochs(params, projs, n_channels, n_epochs, scalings,
         raise RuntimeError('Some channels not classified. Please'
                            ' check your picks')
     ch_names = [params['info']['ch_names'][idx] for idx in inds]
+    _validate_type(params['colors'], (list, None), 'colors')
+    if params['colors'] is not None:
+        if len(params['colors']) != len(params['epochs'].events):
+            raise ValueError('colors must be list of len(epochs.events)'
```
Comment (agramfort) [inline mne/viz/epochs.py]:
```suggestion
            raise ValueError('colors must be list of len(epochs.events). '
```

File: mne/viz/epochs.py [hunk 7]
```diff
@@ -978,6 +982,22 @@ def _prepare_mne_browse_epochs(params, projs, n_channels, n_epochs, scalings,
...
+                             'Got %s' % len(params['colors']))
+        for epoch_idx in range(len(params['colors'])):
+            these_colors = params['colors'][epoch_idx]
+            _validate_type(these_colors, list,
+                           'color[%s]' % (epoch_idx,))
+            if len(these_colors) != n_channels:
+                raise ValueError('colors for the %dth epoch'
```
Comment (agramfort) [inline mne/viz/epochs.py]:
```suggestion
                raise ValueError('colors for the %dth epoch '
```

File: mne/viz/epochs.py [hunk 8]
```diff
@@ -978,6 +982,22 @@ def _prepare_mne_browse_epochs(params, projs, n_channels, n_epochs, scalings,
...
+                                 'has length %d, expected %d'
```
Comment (agramfort) [inline mne/viz/epochs.py]:
```suggestion
                                 'has length %d, expected %d.'
```

**Comment by jasmainak** [comment]
@drammock we agreed with @larsoner that for now we don't expose this in a tutorial or example. The main purpose of this PR is to fix autoreject which fails now with the latest MNE release.

If we want to expose this in a tutorial, the correct approach would be to allow also a flat list, not just a nested list of list as it's more complicated to demo in an example. Then you would have to implement broadcasting and the code begins to get more complicated. Probably not worth the effort and future complexity in maintaining the code

**Comment by jasmainak** [comment]
CIs are happy

**Comment by agramfort** [comment]
Thx @jasmainak

**Pull request merged**